src/execution/window_executor.cpp

```diff
@@ -602,6 +605,143 @@ void WindowExecutorBoundsState::UpdateBounds(idx_t row_idx, DataChunk &input_chu
+class ExclusionFilter {
+public:
+	ExclusionFilter(WindowExclusion exclude_mode_p, idx_t total_count, const ValidityMask &src) : mode(exclude_mode_p) {
+		mask.Initialize(total_count);
+
+		// initialize mask_src:
+		if (src.GetData()) {
+			mask_src = &src;
+		} else {
+			// if src has no data (= represents only one entries), set the mask source to all_ones_mask)
+			one_bits.resize(ValidityMask::ValidityMaskSize(total_count), ~0);
+			all_ones_mask.Initialize(one_bits.data());
+			mask_src = &all_ones_mask;
+		}
+		// copy the data from mask_src
+		FetchFromSource(0, total_count - 1);
+	}
+
+	//! Copy the entries from mask_src to mask, in the index range [begin, end)
+	void FetchFromSource(idx_t begin, idx_t end);
+	//! Apply the current exclusion to the validity mask
+	//! (offset is the current row's index within the chunk)
+	void ApplyExclusion(DataChunk &bounds, idx_t row_idx, idx_t offset);
+	//! Reset the validity mask to match mask_src
+	//! (offset is the current row's index within the chunk)
+	void ResetMask(idx_t row_idx, idx_t offset);
+
+	//! The current peer group's begin
+	idx_t curr_peer_begin;
+	//! The current peer group's end
+	idx_t curr_peer_end;
+	//! The window exclusion mode
+	WindowExclusion mode;
+	//! The validity mask representing the exclusion
+	ValidityMask mask;
+	//! The validity mask upon which mask is based
+	const ValidityMask *mask_src;
+	//! A validity mask consisting of only one entries (needed if no ignore_nulls mask is supplied)
+	ValidityMask all_ones_mask;
+	//! A field of one bits
+	vector<validity_t> one_bits;
+};
+
+void ExclusionFilter::FetchFromSource(idx_t begin, idx_t end) {
+	idx_t begin_entry_idx, end_entry_idx, buf;
+	mask.GetEntryIndex(begin, begin_entry_idx, buf);
+	mask.GetEntryIndex(end, end_entry_idx, buf);
+	auto dst = mask.GetData() + begin_entry_idx;
+	auto src = mask_src->GetData() + begin_entry_idx;
+	for (idx_t entry_idx = begin_entry_idx; entry_idx++ <= end_entry_idx;) {
+		*dst++ = *src++;
+	}
+}
+
+void ExclusionFilter::ApplyExclusion(DataChunk &bounds, idx_t row_idx, idx_t offset) {
+	// flip the bits in mask according to the window exclusion mode
+	switch (mode) {
+	case WindowExclusion::CURRENT_ROW:
+		mask.SetInvalid(row_idx);
+		break;
+	case WindowExclusion::TIES:
+	case WindowExclusion::GROUP: {
+		if (curr_peer_end == row_idx || offset == 0) {
+			// new peer group or input chunk: set entire peer group to invalid
+			auto peer_begin = FlatVector::GetData<const idx_t>(bounds.data[PEER_BEGIN]);
+			auto peer_end = FlatVector::GetData<const idx_t>(bounds.data[PEER_END]);
+			curr_peer_begin = peer_begin[offset];
+			curr_peer_end = peer_end[offset];
+			for (idx_t i = curr_peer_begin; i < curr_peer_end; i++) {
+				mask.SetInvalid(i);
+			}
+		}
+		if (mode == WindowExclusion::TIES) {
+			mask.Set(row_idx, mask_src->RowIsValid(row_idx));
+		}
+		break;
+	}
+	default:
+		break;
+	}
+}
+
+void ExclusionFilter::ResetMask(idx_t row_idx, idx_t offset) {
```

Comment:
Is this for reusability in other windows?

Reply:
No I believe it is for the "Value" executors.

Reply:
> No I believe it is for the "Value" executors.

Exactly.